src/Microsoft.Diagnostics.NETCore.Client/ReversedServer/ReversedDiagnosticsServer.cs

```diff
@@ -332,24 +387,41 @@ private void ProvideStream(Guid runtimeId, Stream stream)
 
         private void RunStreamHandlers(Guid runtimeId, Stream stream)
         {
-            Debug.Assert(Monitor.IsEntered(_lock));
+            Debug.Assert(Monitor.IsEntered(_streamLock));
 
             // If there are any handlers waiting for a stream, provide
             // it to the first handler in the queue.
-            for (int i = 0; (i < _handlers.Count) && (null != stream); i++)
+            bool consumedStream = false;
+            for (int i = 0; !consumedStream && i < _streamHandlers.Count; i++)
             {
-                StreamHandler handler = _handlers[i];
-                if (handler(runtimeId, ref stream))
+                StreamHandler handler = _streamHandlers[i];
+                if (handler(runtimeId, stream, out consumedStream))
                 {
-                    _handlers.RemoveAt(i);
+                    _streamHandlers.RemoveAt(i);
                     i--;
                 }
             }
 
-            // Store the stream for when a handler registers later. If
-            // a handler already captured the stream, this will be null, thus
-            // representing that no existing stream is waiting to be consumed.
-            _cachedStreams[runtimeId] = stream;
+            // Store the stream for when a handler registers later.
+            _cachedStreams[runtimeId] = consumedStream ? null : stream;
+        }
+
+        private void RegisterStreamHandler(Guid runtimeId, StreamHandler handler)
+        {
+            lock (_streamLock)
+            {
+                if (!_cachedStreams.TryGetValue(runtimeId, out Stream stream))
+                {
+                    throw new InvalidOperationException($"Runtime instance with identifier '{runtimeId}' is not registered.");
```

Comment:
Setting the key's value to null doesn't remove it from the dictionary, it just makes the value null. So when the second Connect comes through, it retrieves the the value (which is null) for the key and skips running the handlers at that time.

Reply:
Arg, aparently clicking unresolve conversation clears all text in the comment box. Oh well...

I forgot that Dictionary preserved nulls as entries. I still think throwing this exception isn't the right behavior but the scenario gets much more obscure for where it makes a difference. If the user had an out-of-band mechanism to discover the runtime cookie it should be legal for them to write code like this:
````C#
var endpoint = new ServerIpcEndpoint(listener, runtimeCookie);
var client = new DiagnosticClient(endpoint);
client.GetProcessInfo();
````
Currently it appears this would throw if a runtime with runtimeCookie hasn't connected yet. Instead correct behavior would be to wait for a connection, similar to other endpoints. Eventually either a connection would be received or the request would timeout.

Reply:
Okay, I understand the motivation now. I was going under the assumption that ServerIpcEndpoint and PidIpcEndpoint would not be public at any time since they are implementation details thus we can control whatever behavior they have. But I'll update the code to optionally get the stream rather than requiring it to have been registered.